jupyter-lib/api/src/main/kotlin/org/jetbrains/kotlinx/jupyter/api/VariableState.kt

```diff
@@ -16,18 +18,81 @@ data class VariableStateImpl(
     override val scriptInstance: Any,
 ) : VariableState {
     private var cachedValue: Any? = null
+    private var isRecursive: Boolean = false
 
     fun update() {
         val wasAccessible = property.isAccessible
         property.isAccessible = true
         val fieldValue = property.get(scriptInstance)
         property.isAccessible = wasAccessible
         cachedValue = fieldValue
+        handleRecursiveCase()
     }
 
     override val stringValue: String?
-        get() = cachedValue?.toString()
+        get() = getProperString()
 
     override val value: Any?
         get() = cachedValue
+
+    private fun handleRecursiveCase() {
+        if (cachedValue == null) return
+        val seenNodes = mutableSetOf<Any>()
+        goThrough(cachedValue, seenNodes)
+    }
+
+    private fun getProperString(): String? {
```

Comment:
It's not necessary to move this logic to a separate function, you can leave it in the getter

Reply:
Indeed. didn't know about it